### ISSUE TITLE
Sonarqube Tests finding mitigations after running Reimport-scan

### ISSUE THREAD

**Issue by Homopatrol** [DefectDojo/django-DefectDojo#3675]
**Bug description**
I’ve come across a bug that causes DefectDojo to erroneously flag mitigations when reuploading a Sonarqube test scan. 

**Steps to reproduce**
I imported a [sonar-report.txt](https://github.com/DefectDojo/defectdojo_api/files/5849498/sonar-report.txt) to DefectDojo

<img width="1659" alt="Screenshot 2021-01-21 at 12 01 52" src="https://user-images.githubusercontent.com/58219367/105355127-86159100-5be9-11eb-897d-cf8e7d042efe.png">

As you can see in the report some errors reported twice however these relate to different components in the project – eg. 
```
"<!DOCTYPE>" Declarations Should Appear Before "<html>" Tags
```
This initial upload provides the following results:

<img width="1032" alt="Screenshot 2021-01-21 at 12 03 31" src="https://user-images.githubusercontent.com/58219367/105353309-fbcc2d80-5be6-11eb-952d-350e15271b97.png">

However, upon reimporting the same report by running the script again DefectDojo seems to find and mitigate 5 results despite the fact I have not changed the report. 

<img width="1028" alt="Screenshot 2021-01-21 at 12 05 44" src="https://user-images.githubusercontent.com/58219367/105353320-fe2e8780-5be6-11eb-84d8-512a956b9da3.png">

**Expected behavior**
I expected the reuploaded test to display the same results with no Mitgated findings.

This error also appears when reuploading Dependency Track and Dependency Check scans so is there an issue with the way DefectDojo is analysing the test results?

**Deployment method** *(select with an `X`)*
- [X] k8s/OpenShift with Helm chart
- [ ] Kubernetes
- [ ] Docker
- [ ] setup.bash / legacy-setup.bash

**Environment information**
N/A

**Sample scan files** (optional)
 [sonar-report.txt](https://github.com/DefectDojo/defectdojo_api/files/5849498/sonar-report.txt)

**Comment by Homopatrol** [comment]
@valentijnscholten Ah okay so this is known to happen with many other scans to. Do you know which part of the code that contains the import/repimport functionality so I could look futher into this issue?

**Comment by valentijnscholten** [comment]
dojo/api_v2/serializers.py and dojo/test/views.py, see also https://github.com/DefectDojo/django-DefectDojo/pull/3629

**Comment by Homopatrol** [comment]
@valentijnscholten Thanks that's very helpful. - Is this being developed futher to fix the other scanners by adding the correct parsers for each or is there a bigger work required?

**Comment by ptrovatelli** [comment]
Hello, 
The re-import matching code is very basic. it didn't evolve when the configurable deduplication was added (my bad) and it wasn't event coherent with legacy deduplication algorithm in the first place
- legacy dedupe basically identifies findings by (see comments in settings.dist.py; i say "basically" because there are some more subtle conditions): 
  - static scanner:  ['title', 'cwe', 'line', 'file_path', 'description']
  -  dynamic scanner: ['title', 'cwe', 'line', 'file_path', 'description', 'endpoints']
- re-import identifies findings by : title, severity, numerical_severity (except a few parsers for which exceptions were hardcoded)
- configurable deduplication (as detailed in the "readthedoc") : 
  - as it's name stands can be tuned as wanted through the hash_code which is itself completely configurable. 
  - Or it can use the "unique_id_from_tool" field in place of the hash_code. 
  - It's also possible to deduplicate on both the hash_code (functional identifier) and the unique_id_from_tool (technical identifier) in order to have find both technically or functionally similar findings

so I guess at this point if one wants a coherent re-import, they would need to configure the deduplication to match the re-import logic: title, severity, numerical_severity (except for the parsers that have changed this by code: Veracode, Arachni, Anchore). That is not ideal though. 

I'll see if I can apply the deduplication configuration to the re-import feature to avoid those discrepancies

**Comment by ptrovatelli** [comment]
@Homopatrol can you re-post your sonar report please? the one provided has no findings. 
Thanks

**Comment by damiencarol** [comment]
Assigned this issue to @ptrovatelli as discussed in #2030

**Comment by ptrovatelli** [comment]
I've started working on something. still some tests to do and unit tests to update, but it should already fix most of the cases : https://github.com/DefectDojo/django-DefectDojo/pull/3753

**Comment by Homopatrol** [comment]
@ptrovatelli Yep sorry this one should be populated. - [sonar-report.txt](https://github.com/DefectDojo/django-DefectDojo/files/5903076/sonar-small-report.txt)

**Comment by valentijnscholten** [comment]
@Homopatrol Can you try again with 2.0.0 or a version that includes #3753? It should be a lot better now.

**Comment by Homopatrol** [comment]
@valentijnscholten

Thanks, this does work now when uploading a Sonar test in v2.0.3 

<img width="959" alt="Screenshot 2021-07-08 at 11 04 19" src="https://user-images.githubusercontent.com/58219367/124907905-99be8380-dfe0-11eb-896e-766942939018.png">

<img width="1448" alt="Screenshot 2021-07-08 at 11 39 55" src="https://user-images.githubusercontent.com/58219367/124908500-50226880-dfe1-11eb-8d84-c2b297f7a6b6.png">

and re-uploading to v2.0.3 👍  

However, I noticed when upgrading to v2.0.3 that a report I had previously uploaded in v1.14.0 was now flagging up mitigations and additional findings after being re-uploaded in v2.0.3.

<img width="455" alt="image" src="https://user-images.githubusercontent.com/58219367/124907683-5a903280-dfe0-11eb-99c8-640bb3d5c28f.png">

<img width="506" alt="image" src="https://user-images.githubusercontent.com/58219367/124907706-62e86d80-dfe0-11eb-8566-ff79189f57e3.png">

Is this expected behavior when upgrading DefectDojo? or is there a way of ensuring integrity with reports and findings when upgrading versions?

**Comment by valentijnscholten** [comment]
Did you perform all the upgrade steps including recalculating all hash codes?

**Comment by Homopatrol** [comment]
Thank you, I ran `./manage.py dedupe --hash_code_only` In my initializer pod and corrected the issue. 

Following this I have amended my initializer-job.yaml so the initializer will perform this script every time there is an upgrade
```
      - name: initializer
        image: "{{ template "initializer.repository" . }}:{{ .Values.tag }}"
        imagePullPolicy: {{ .Values.imagePullPolicy }}
        {{- if .Values.securityContext.enabled }}
        securityContext:
          {{- toYaml .Values.securityContext.djangoSecurityContext | nindent 10 }}
        {{- end }}          
        command: ["/bin/sh", "-c"]
        args:
        {{- if .Release.IsUpgrade }}
          - >
            /entrypoint-initializer.sh &&
            python manage.py dedupe --hash_code_only
        {{ else }}
          - /entrypoint-initializer.sh
        {{- end }}
```
Is there a reason the current initializer-job.yaml does not perform this `manage.py dedupe --hash_code_only` script on every upgrade, if not could you implement this feature into the upstream initializer-job.yaml?

**Comment by valentijnscholten** [comment]
We discussed that but didn't make a decision. On larger installs it can run for a couple of hours, so it's easier to let the admin control it. Unless the admin doesn't read the upgrade instructions ... We could make it async maybe.
Either way, the initializer script itself should be changed, not the k8s probably.
Where is the {{.Release.IsUpgrade}} coming from in your example?

**Comment by mtcolman** [comment]
@valentijnscholten `.Release.IsUpgrade` is a built-in object in helm: https://helm.sh/docs/chart_template_guide/builtin_objects/
>    Release.Name: The release name
    Release.Namespace: The namespace to be released into (if the manifest doesn’t override)
    Release.IsUpgrade: This is set to true if the current operation is an upgrade or rollback.
    Release.IsInstall: This is set to true if the current operation is an install.
    Release.Revision: The revision number for this release. On install, this is 1, and it is incremented with each upgrade and rollback.
    Release.Service: The service that is rendering the present template. On Helm, this is always Helm.

**Comment by ptrovatelli** [comment]
@Homopatrol great! how about creating a PR with that evolution? to prevent unexpected recomputations, I think we should also add a parameter in the initializer's configuration (values.yaml) like "recomputeHashcodesUponUpgrade" defaulted to false so that ppl can activate it or not willingly

 @valentijnscholten your thoughts? 

thanks!

**Comment by valentijnscholten** [comment]
Good idea. Would be nice to have something similar in the docker-compose setup. But I don't think it has an `IsUpgrade` boolean built in, so it might require custom code in the initializer.

**Comment by ptrovatelli** [comment]
the problem with changing the initalizer script only is that it won't know whether we're doing an upgrade or not will it? or maybe automatically recompute if we have detected a model change. that would be better that being based on Release.IsUpgrade because it will trigger less often (won't trigger for upgrades that don't include a model change). 

how about opening an issue first to discuss solutions first?

**Comment by manuel-sommer** [comment]
Is this issue still present? I would suggest to close this issue as stale @valentijnscholten 
The parser Sonarqube has also changed since then. If there are still issues left, I would suggest to open a new issue.
